### PR TITLE
Non-native QML camera enhancements

### PR DESCRIPTION
**Pull request by nirvn**
Enhancements include:
- significant UI/UX improvements
- zoom controls (pinch and mouse wheel, with button to reset to 1.0x zoom)
- flash controls
- GPS metadata
- video recording duration label

Still image capture:
image

Video capture:
image

The PR also fixes a longstanding issue whereas resizing photos to match the qfieldsync maximum image size setting would result in loss of metadata.

### CODE REVIEW

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 9b87dbb98f981de930f567f306ee6d4dcd86b923: [**arm64-android**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-3999-8773cb8fc714149274323b4227db8ae18cd23940-arm64-android.apk)
<details>
<summary>Other architectures</summary>

- [arm-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-3999-8773cb8fc714149274323b4227db8ae18cd23940-arm-android.apk)
- [x64-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-3999-8773cb8fc714149274323b4227db8ae18cd23940-x64-android.apk)
- [x86-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-3999-8773cb8fc714149274323b4227db8ae18cd23940-x86-android.apk)
- [**all access arm64-android**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield_all_access-dev-3999-8773cb8fc714149274323b4227db8ae18cd23940-arm64-android.apk)
- [all access arm-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield_all_access-dev-3999-8773cb8fc714149274323b4227db8ae18cd23940-arm-android.apk)
- [all access x64-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield_all_access-dev-3999-8773cb8fc714149274323b4227db8ae18cd23940-x64-android.apk)
- [all access x86-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield_all_access-dev-3999-8773cb8fc714149274323b4227db8ae18cd23940-x86-android.apk)
</details>

File: src/qml/QFieldCamera.qml [hunk 1]
```diff
@@ -74,62 +132,60 @@ Popup {
...
-        bgcolor: "grey"
-        borderColor: Theme.mainColor
+      property real oldScale: 1.0
 
-        onClicked: camera.imageCapture.captureToLocation(qgisProject.homePath+ '/DCIM/')
+      onActiveChanged: {
+        if (active) {
+          oldScale = 1.0
+        } else {
+        }
```
Comment (suricactus) [inline src/qml/QFieldCamera.qml:159]: Else?

File: src/qml/QFieldCamera.qml [hunk 2]
```diff
@@ -74,62 +132,60 @@ Popup {
...
+      onWheel: (event) => {
+                 if (event.angleDelta.y > 0)
+                 {
+                   camera.zoomIn(0.25)
+                 }
+                 else
+                 {
+                   camera.zoomOut(0.25)
+                 }
+               }
```
Comment (suricactus) [inline src/qml/QFieldCamera.qml:186]: Indent

File: src/qml/QFieldCamera.qml [hunk 3]
```diff
@@ -156,38 +212,176 @@ Popup {
...
+
+      Rectangle {
+        id: captureRing
+        anchors.centerIn: parent
+        width: 64
+        height: 64
+        radius: 32
+        color: Qt.hsla(Theme.darkGray.hslHue, Theme.darkGray.hslSaturation, Theme.darkGray.hslLightness, 0.3)
+        border.color: cameraItem.state == "VideoCapture" && camera.videoRecorder.recorderState != CameraRecorder.StoppedState
+                      ? "red" : "white"
```
Comment (suricactus) [inline src/qml/QFieldCamera.qml:231]:
```suggestion
                      ? "red"
                      : "white"
```

Lets have it consistent, either 1 when condition is short or 3 lines when too long.

File: src/qml/QFieldCamera.qml [hunk 4]
```diff
@@ -156,38 +212,176 @@ Popup {
...
+          }
+        }
+      }
 
-      onClicked: cameraItem.finished(currentPath)
+      Rectangle {
+        visible: cameraItem.state == "VideoCapture" && camera.videoRecorder.recorderState != CameraRecorder.StoppedState
+
+        x: cameraItem.isPortraitMode ? captureRing.x + captureRing.width / 2 - width / 2 : captureRing.x + captureRing.width / 2 - width / 2
+        y: cameraItem.isPortraitMode ? captureRing.y - height - 20 : captureRing.y - height - 20
```
Comment (suricactus) [inline src/qml/QFieldCamera.qml:328]: I guess you know you are the only person in the universe understanding this `x` and `y` calculation.

Any chance that we have a "toolbar" where they are rendered sequentially? Or any kind of function calculating this?

**Review by suricactus — APPROVED** [review]
Tried with photo camera without video.

File: src/qml/QFieldCamera.qml (hunk 1 above)
Comment (nirvn) [reply src/qml/QFieldCamera.qml:159]: @suricactus , else you get nothing :wink:

File: src/qml/QFieldCamera.qml (hunk 3 above)
Comment (nirvn) [reply src/qml/QFieldCamera.qml:231]: Agreed.

File: src/qml/QFieldCamera.qml (hunk 4 above)
Comment (nirvn) [reply src/qml/QFieldCamera.qml:328]: @suricactus , unfortunately not sir; complexity comes from the fact that we want the controls to be sitting at the bottom of the screen when in portrait and on the right edge of the screen when in landscape.

**Pull request merged**